game/src/main/scala/io/aigar/game/Virus.scala

```diff
@@ -22,35 +22,19 @@ class Virus(spawnPosition: Vector2) extends Entity {
 class Viruses(grid: Grid) extends EntityContainer {
   var viruses = List.fill(Virus.Quantity)(new Virus(grid.randomPosition))
 
-  def update(grid: Grid, players: List[Player]): Unit = {
-
-    detectCollisions(players)
+  def shouldRespawn: Boolean = viruses.size < Virus.Quantity
 
-    if(viruses.size < Virus.Quantity) {
-      val position = respawn(grid, players, Virus.RespawnRetryAttempts)
+  def update(grid: Grid, players: List[Player]): Unit = {
```

Comment:
Could this be in `EntityContainer`? I don't know if it would be cleaner, but I think that you'll do many copy/paste of this code.